JS: whitelilist delimiter unwrapping for js/incomplete-sanitization

esben-semmle opened the pull request:
Addresses a part of <https://jira.semmle.com/browse/ODASA-7891>. Unwrapper-calls like `var unwrapped = str.replace('{', '').replace('}', '')` are no longer flagged.

Evaluation is ongoing. This is being considered for the 1.20.1 patch release.

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/tst.js:
```diff
@@ -126,6 +126,21 @@ function good11(s) {
+
+	s = s.replace('[', ''); // OK, but still flagged
+	s = s.replace(']', ''); // OK, but still flagged
+	s.replace(/{/, '').replace('/}/', ''); // OK, but still flagged
```
Why is this OK? The quotes around the second regex look pretty un-OK to me.

xiemaisi commented on javascript/ql/src/Security/CWE-116/IncompleteSanitization.ql:
```diff
@@ -101,6 +101,40 @@ predicate allBackslashesEscaped(DataFlow::Node nd) {
+ */
+predicate removesFirstOccurence(MethodCallExpr repl, string str) {
+  repl.getMethodName() = "replace" and
+  repl.getArgument(0).getStringValue() = str and
```
Use `getAMatchedString` here to handle the regex case?

esben-semmle replied:
That is a deliberate limitation. I think it is more likely that the global flag was forgotten if this is a regex case, and I can not recall any FPs due do regexes.

xiemaisi replied:
OK, makes sense.

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/tst.js:
```diff
@@ -126,6 +126,21 @@ function good11(s) {
+	s.replace('(', '').replace(')', ''); // OK
+	s.replace('{', '').replace('}', ''); // OK
+
+	s.replace('[', '\\[').replace(']', '\\]'); // probably OK, but still flagged
```
I don't quite see why this would be "probably OK".

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/tst.js:
```diff
@@ -126,6 +126,21 @@ function good11(s) {
+	s.replace('{', '').replace('}', ''); // OK
+
+	s.replace('[', '\\[').replace(']', '\\]'); // probably OK, but still flagged
+	s.replace('(', '\\[').replace(')', '\\)'); // OK
```
What is this testing? Is the replacement of `(` with `\\[` a copy-paste error?

xiemaisi commented on javascript/ql/test/query-tests/Security/CWE-116/IncompleteSanitization/tst.js:
```diff
@@ -126,6 +126,21 @@ function good11(s) {
+	s.replace('{', '\\{').replace('}', '\\}'); // probably OK, but still flagged
+
+	s = s.replace('[', ''); // OK, but still flagged
+	s = s.replace(']', ''); // OK, but still flagged
```
Presumably you could handle this by using local data flow in `isDelimiterUnwrapper`?

xiemaisi commented:
One more thing: currently this is targeted at `master`; I assume you have checked that it could cleanly be retargeted at `rc/1.20` for inclusion in 1.20.1?

esben-semmle commented:
Some changes motivated by the evaluation and the review:

- cleaned up the tests, and made my intentions clearer. 
- added local dataflow (I was worried about local data flow leading to false negatives, but that is not a problem in practice)
- removed `<`/`>` as a whitelisted delimiter pair, it seems too likely to be a case of bad HTML escaping
- retargeted 1.20

xiemaisi commented on change-notes/1.20/analysis-javascript.md:
```diff
@@ -43,6 +43,7 @@
 | Assignment to exports variable             | Fewer results                | This query no longer flags code that is also flagged by the query "Useless assignment to local variable". |
 | Client-side cross-site scripting           | More true positive and fewer false positive results. | This query now recognizes WinJS functions that are vulnerable to HTML injection. It no longer flags certain safe uses of jQuery, and recognizes custom sanitizers. |
 | Hard-coded credentials                     | Fewer false positive results | This query no longer flags the empty string as a hardcoded username. |
+| Incomplete string escaping or encoding | Fewer false-positive results | This rule no longer flags calls that unwraps a string from a pair of delimiters. |
```
Change notes for patch releases don't go here, I think; this should go on the JIRA issue instead.

Also: "unwraps" is singular, but "calls" is plural.

esben-semmle replied:
Fixed. Added change note to specialized ticket: https://jira.semmle.com/browse/ODASA-7907.
